src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/AbstractConditionMapper.java

```diff
@@ -57,28 +59,46 @@
+        // validate no table primary key field is being updated, and that if one field in a secondary index is
+        // being updated, then the other is as well
+        validateFieldsCanBeUpdated(actions.keySet());
+
+        // map virtual field values to physical ones
+        Map<String, AttributeValue> physicalActions = itemMapper.applyForWrite(actions);
+
+        // construct physical update SET expression
+        List<String> clauses = physicalActions.entrySet().stream()
+            .map(entry -> getUpdateClause(request, entry.getKey(), entry.getValue(), delimiter))
+            .collect(Collectors.toList());
+
+        return clauses;
+    }
+
     @Override
     public void applyForUpdate(UpdateItemRequest updateItemRequest) {
         if (updateItemRequest.getUpdateExpression() != null) {
             RequestWrapper request = new UpdateExpressionRequestWrapper(updateItemRequest);
 
             // parse update expression for the field values being set
-            Map<String, AttributeValue> virtualSetActions = parseUpdateExpression(request,
-                updateItemRequest.getConditionExpression());
-            checkArgument(!virtualSetActions.isEmpty(), "Update expression needs at least one SET action");
-
-            // validate no table primary key field is being updated, and that if one field in a secondary index is
-            // being updated, then the other is as well
-            validateFieldsCanBeUpdated(virtualSetActions.keySet());
-
-            // map virtual field values to physical ones
-            Map<String, AttributeValue> physicalSetActions = itemMapper.applyForWrite(virtualSetActions);
-
-            // construct physical update SET expression
-            List<String> clauses = physicalSetActions.entrySet().stream()
-                .map(entry -> getUpdateSetClause(request, entry.getKey(), entry.getValue()))
-                .collect(Collectors.toList());
-            request.setExpression("SET " + String.join(", ", clauses));
+            Pair<Map<String, AttributeValue>, Map<String, AttributeValue>> virtualActions = parseUpdateExpression(
+                request, updateItemRequest.getConditionExpression());
+            Map<String, AttributeValue> virtualSetActions = virtualActions.getFirst();
+            Map<String, AttributeValue> virtualAddActions = virtualActions.getSecond();
+
+            checkArgument(!virtualSetActions.isEmpty() || !virtualAddActions.isEmpty(),
+                "Update expression needs at least one supported action: SET or ADD");
+
+            List<String> setClauses = getPhysicalUpdateExpressionForAction(request, virtualSetActions, " = ");
+            List<String> addClauses = getPhysicalUpdateExpressionForAction(request, virtualAddActions, " ");
+
+            String setExpression = setClauses.isEmpty() ? "" : "SET " + String.join(", ", setClauses);
+            String addExpression = addClauses.isEmpty() ? "" : "ADD " + String.join(", ", addClauses);
```

Comment:
I don't know if this will work with hash partitioning, if we're updating a field in a secondary index. If it's an index field, then the virtual numerical attribute that we're adding to will have its value encoded in a binary index attribute, so we can't just do a straight ADD.

Reply:
I see, that's probably why there is validation for SET in Hash Partitioning?

Reply:
Where is this validation?

Reply:
https://github.com/salesforce/mt-dynamo/blob/27e9bde11e1fc7fe4ad99eca9d688bc45de26900/src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/HashPartitioningConditionMapper.java#L74